File: src/main/java/com/example/jutil/DataBlock.java
package com.example.jutil;

import java.util.*;

/**
 * Created by yilong on 2018/2/22.
 */
public class DataBlock {
    public List<byte[]> buf = new ArrayList<byte[]>();
    public int cur = 0;
    public String lastKey;
    public String firstKey;

    public DataBlock(String firstKey) {
        this.firstKey = firstKey;
        this.lastKey = firstKey;
    }

    public boolean save(byte[] data) {
        if (Constant.BLOCK_SIZE < data.length + cur) {
            return false;
        }

        buf.add(data);
        cur += data.length;

        return true;
    }

    public int avaliableSize() {
        return (Constant.BLOCK_SIZE - cur);
    }

    public void updateLastKey(String k) {
        if (k.compareTo(lastKey) > 0) {
            lastKey = k;
        }
    }

    public String getLastKey() {
        return lastKey;
    }

    public List<byte[]>  getBufArray() {
        return buf;
    }

    public void  clean() {
        buf.clear();
    }
}

File: src/main/scala/com/example/sutil/DataBlock.scala
package com.example.sutil

import com.example.jutil.Constant

import scala.collection._

/**
  * Created by yilong on 2018/1/25.
  */
class DataBlock(var firstKey : String) extends IBlock {
  //val buf = java.nio.ByteBuffer.allocate(Constant.BLOCK_SIZE)
  private val buf = new mutable.ListBuffer[Array[Byte]]
  private var cur = 0
  var lastKey = firstKey

  def save(data : Array[Byte]) : Boolean = {
    (Constant.BLOCK_SIZE >= data.length + cur) match {
      case false => false
      case true => {
        //System.arraycopy(data,0,buf,cur,data.length)
        //buf.put(data, cur, data.length)
        buf.append(data)
        cur += data.length
        true
      }
    }
  }

  def avaliableSize() : Int = {
    Constant.BLOCK_SIZE - cur
  }

  //def getBuf() : Array[Byte] = {
  //  buf.array()
  //}

  def updateLastKey(k:String) = {
    lastKey = k
  }

  def getLastKey() : String = {
    lastKey
  }

  def getBufArray() : mutable.ListBuffer[Array[Byte]] = {
    buf
  }

  def clean() = {
    buf.clear()
  }
}
